scipy/sparse/_index.py

```diff
@@ -322,6 +324,20 @@ def _check_ellipsis(index):
     return index[:first_ellipsis] + (slice(None),)*nslice + tuple(tail)
 
 
+def _compatible_boolean_index(idx):
+    """Returns a boolean index array that can be converted to
+    integer array. Returns None if no such array exists """
+    if hasattr(idx, '__iter__'):
```

Comment:
This seems a bit too permissive. For instance, a dictionary with boolean keys would be an accepted index.

How about:

```python
if not hasattr(idx, 'ndim') or not hasattr(idx, 'dtype'):
    idx = np.asanyarray(idx)
if idx.dtype.kind == 'b':
    return idx
return None
```

Reply:
Yes, this seems to simpler and better. I will make the changes.